In [1]:
from __future__ import print_function
import numpy as np
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
from torchvision import datasets, transforms
# from torch.utils.data import DataLoader, random_split

In [2]:
class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        
        self.conv1 = nn.Sequential(
            nn.Conv2d(1, 16, 3, padding=1),
            nn.ReLU(),
            nn.BatchNorm2d(16),
            nn.Conv2d(16, 16, 3, stride=2, padding=1),
            nn.ReLU(),
            nn.BatchNorm2d(16),
            nn.MaxPool2d(2, 2),
            nn.Dropout(0.25)
        )
        
        self.conv2 = nn.Sequential(
            nn.Conv2d(16, 16, 3, padding=1),
            nn.ReLU(),
            nn.BatchNorm2d(16),
            nn.Conv2d(16, 16, 3, stride=2, padding=1),
            nn.ReLU(),
            nn.BatchNorm2d(16),
            nn.MaxPool2d(2, 2),
            nn.Dropout(0.25)
        )

        self.conv3 = nn.Sequential(
            nn.Conv2d(16, 16, 3, padding=1),
            nn.ReLU(),
            nn.BatchNorm2d(16),
            nn.Conv2d(16, 16, 3, stride=2, padding=1),
            nn.ReLU(),
            nn.BatchNorm2d(16),
            nn.Dropout(0.25)
        )
        
        
    def forward(self, x):
        x = self.conv1(x)
        x = self.conv2(x)
        x = self.conv3(x)
        
        x = x.view(x.size(0), -1)
        # x = nn.GovalAvgPooling(x)
        x = F.log_softmax(x, dim=1)
        return x

In [3]:
!pip install torchsummary
from torchsummary import summary
use_cuda = torch.cuda.is_available()
device = torch.device("cuda" if use_cuda else "cpu")
model = Net().to(device)
summary(model, input_size=(1, 28, 28))

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
            Conv2d-1           [-1, 16, 28, 28]             160
              ReLU-2           [-1, 16, 28, 28]               0
       BatchNorm2d-3           [-1, 16, 28, 28]              32
            Conv2d-4           [-1, 16, 14, 14]           2,320
              ReLU-5           [-1, 16, 14, 14]               0
       BatchNorm2d-6           [-1, 16, 14, 14]              32
         MaxPool2d-7             [-1, 16, 7, 7]               0
           Dropout-8             [-1, 16, 7, 7]               0
            Conv2d-9             [-1, 16, 7, 7]           2,320
             ReLU-10             [-1, 16, 7, 7]               0
      BatchNorm2d-11             [-1, 16, 7, 7]              32
           Conv2d-12             [-1, 16, 4, 4]           2,320
    

In [4]:


torch.manual_seed(1)
batch_size = 128

kwargs = {'num_workers': 1, 'pin_memory': True} if use_cuda else {}
train_loader = torch.utils.data.DataLoader(
    datasets.MNIST('../data', train=True, download=True,
                    transform=transforms.Compose([
                        transforms.ToTensor(),
                        transforms.Normalize((0.1307,), (0.3081,))
                    ])),
    batch_size=batch_size, shuffle=True, **kwargs)
test_loader = torch.utils.data.DataLoader(
    datasets.MNIST('../data', train=False, transform=transforms.Compose([
                        transforms.ToTensor(),
                        transforms.Normalize((0.1307,), (0.3081,))
                    ])),
    batch_size=batch_size, shuffle=True, **kwargs)


In [5]:
print("Train datset size =",len(train_loader))
print("Test datset size =",len(test_loader))

Train datset size = 469
Test datset size = 79


In [6]:
from tqdm import tqdm
def train(model, device, train_loader, optimizer, epoch):
    model.train()
    pbar = tqdm(train_loader)
    for batch_idx, (data, target) in enumerate(pbar):
        data, target = data.to(device), target.to(device)
        optimizer.zero_grad()
        output = model(data)
        loss = F.nll_loss(output, target)
        loss.backward()
        optimizer.step()
        pbar.set_description(desc= f'loss={loss.item()} batch_id={batch_idx}')


def test(model, device, test_loader):
    model.eval()
    test_loss = 0
    correct = 0
    with torch.no_grad():
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)
            test_loss += F.nll_loss(output, target, reduction='sum').item()  # sum up batch loss
            pred = output.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
            correct += pred.eq(target.view_as(pred)).sum().item()

    test_loss /= len(test_loader.dataset)

    print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.4f}%)\n'.format(
        test_loss, correct, len(test_loader.dataset),
        100. * correct / len(test_loader.dataset)))

In [7]:

model = Net().to(device)
optimizer = optim.SGD(model.parameters(), lr=0.01, momentum=0.9)

for epoch in range(1, 20):
    print("################# Epoch =",str(epoch),"#######################")
    train(model, device, train_loader, optimizer, epoch)
    test(model, device, test_loader)

################# Epoch = 1 #######################


loss=0.6686771512031555 batch_id=468: 100%|██████████| 469/469 [00:22<00:00, 20.61it/s]



Test set: Average loss: 0.1489, Accuracy: 9704/10000 (97.0400%)

################# Epoch = 2 #######################


loss=0.730696439743042 batch_id=468: 100%|██████████| 469/469 [00:18<00:00, 24.71it/s]



Test set: Average loss: 0.0900, Accuracy: 9827/10000 (98.2700%)

################# Epoch = 3 #######################


loss=0.48139891028404236 batch_id=468: 100%|██████████| 469/469 [00:18<00:00, 25.56it/s]



Test set: Average loss: 0.0719, Accuracy: 9851/10000 (98.5100%)

################# Epoch = 4 #######################


loss=0.5278867483139038 batch_id=468: 100%|██████████| 469/469 [00:18<00:00, 25.58it/s]



Test set: Average loss: 0.0640, Accuracy: 9848/10000 (98.4800%)

################# Epoch = 5 #######################


loss=0.43508341908454895 batch_id=468: 100%|██████████| 469/469 [00:19<00:00, 23.65it/s]



Test set: Average loss: 0.0571, Accuracy: 9858/10000 (98.5800%)

################# Epoch = 6 #######################


loss=0.4512786865234375 batch_id=468: 100%|██████████| 469/469 [00:18<00:00, 24.75it/s]



Test set: Average loss: 0.0489, Accuracy: 9875/10000 (98.7500%)

################# Epoch = 7 #######################


loss=0.5738426446914673 batch_id=468: 100%|██████████| 469/469 [00:18<00:00, 25.06it/s]



Test set: Average loss: 0.0467, Accuracy: 9884/10000 (98.8400%)

################# Epoch = 8 #######################


loss=0.45721080899238586 batch_id=468: 100%|██████████| 469/469 [00:18<00:00, 24.79it/s]



Test set: Average loss: 0.0420, Accuracy: 9887/10000 (98.8700%)

################# Epoch = 9 #######################


loss=0.4907257556915283 batch_id=468: 100%|██████████| 469/469 [00:19<00:00, 24.34it/s]



Test set: Average loss: 0.0396, Accuracy: 9897/10000 (98.9700%)

################# Epoch = 10 #######################


loss=0.4142589867115021 batch_id=468: 100%|██████████| 469/469 [00:19<00:00, 24.58it/s]



Test set: Average loss: 0.0354, Accuracy: 9904/10000 (99.0400%)

################# Epoch = 11 #######################


loss=0.6172873377799988 batch_id=468: 100%|██████████| 469/469 [00:18<00:00, 25.27it/s]



Test set: Average loss: 0.0348, Accuracy: 9908/10000 (99.0800%)

################# Epoch = 12 #######################


loss=0.431485652923584 batch_id=468: 100%|██████████| 469/469 [00:19<00:00, 24.50it/s]



Test set: Average loss: 0.0334, Accuracy: 9906/10000 (99.0600%)

################# Epoch = 13 #######################


loss=0.5540325045585632 batch_id=468: 100%|██████████| 469/469 [00:19<00:00, 24.39it/s]



Test set: Average loss: 0.0368, Accuracy: 9905/10000 (99.0500%)

################# Epoch = 14 #######################


loss=0.5314993262290955 batch_id=468: 100%|██████████| 469/469 [00:18<00:00, 25.27it/s]



Test set: Average loss: 0.0341, Accuracy: 9905/10000 (99.0500%)

################# Epoch = 15 #######################


loss=0.5220826268196106 batch_id=468: 100%|██████████| 469/469 [00:18<00:00, 25.34it/s]



Test set: Average loss: 0.0353, Accuracy: 9910/10000 (99.1000%)

################# Epoch = 16 #######################


loss=0.42854222655296326 batch_id=468: 100%|██████████| 469/469 [00:19<00:00, 24.32it/s]



Test set: Average loss: 0.0310, Accuracy: 9919/10000 (99.1900%)

################# Epoch = 17 #######################


loss=0.5371251702308655 batch_id=468: 100%|██████████| 469/469 [00:19<00:00, 24.47it/s]



Test set: Average loss: 0.0299, Accuracy: 9917/10000 (99.1700%)

################# Epoch = 18 #######################


loss=0.37640416622161865 batch_id=468: 100%|██████████| 469/469 [00:18<00:00, 25.27it/s]



Test set: Average loss: 0.0341, Accuracy: 9902/10000 (99.0200%)

################# Epoch = 19 #######################


loss=0.5001823306083679 batch_id=468: 100%|██████████| 469/469 [00:18<00:00, 24.94it/s]



Test set: Average loss: 0.0317, Accuracy: 9911/10000 (99.1100%)



## Final Results
Total params: 11,952

Total Epochs: 20

Test set: Average loss: 0.0299, Accuracy: 9917/10000 (99.1700%)

We were able to achine 99.1700% test accuracy on 17th epoch.